image-extension.md

```diff
@@ -0,0 +1,234 @@
+- SHOULD write SBOM (Software-Bill-of-Materials) files as described in the [Software-Bill-of-Materials](#software-bill-of-materials) section describing any contributions to the app image or build environment.
+
+#### Dockerfile Requirements
+
+run.Dockerfiles:
+
+- MAY contain multiple `FROM` instructions
+- MAY contain any instruction except for `CMD` and `ENTRYPOINT`
+
+build.Dockerfiles and Dockerfiles:
+
+- MUST begin with:
+```bash
+ARG base_image
+FROM ${base_image}
+```
+- MUST NOT contain any other `FROM` instructions
+- MAY contain `ADD`, `ARG`, `COPY`, `ENV`, `LABEL`, `RUN`, `SHELL`, `USER`, and `WORKDIR` instructions
+- MUST NOT contain any other instructions
+
+All files:
+
+- SHOULD use the `build_id` build arg to invalidate the cache after a certain layer. When the `$build_id` build arg is referenced in a `RUN` instruction, all subsequent layers will be rebuilt on the next build (as the value will change).
+- SHOULD use the `user_id` and `group_id` build args to reset the image config's `User` field to its original value if the user needs to be changed (e.g., to `root`) during the build in order to perform privileged operations.
+
+## Phase: Extension
+
+### Purpose
+
+The purpose of the extension phase is to apply the Dockerfiles generated in the build-ext phase to the build and/or run base images. The Extension phase MUST NOT be run for Windows builds.
+
+### Process
+
+The extension phase MUST be invoked separately for each base image (build or run) that is extended.
+
+**GIVEN:**
+- The final ordered group of Dockerfiles generated during the build-ext phase,
+- A list of build args for each Dockerfile specified during the build-ext phase,
+
+For each Dockerfile in the group in order, the lifecycle MUST apply the Dockerfile to the base image as follows:
+
+- The lifecycle MUST provide each Dockerfile with:
+- A `base_image` build arg
+    - For the first Dockerfile, the value MUST be the original base image.
+    - When there are multiple Dockerfiles, the value MUST be the intermediate image generated from the application of the previous Dockerfile.
+- A `build_id` build arg
+    - The value MUST be a UUID
+- `user_id` and `group_id` build args
+    - For the first Dockerfile, the values MUST be the original `uid` and `gid` from the `User` field of the config for the original base image.
+    - When there are multiple Dockerfiles, the values MUST be the `uid` and `gid` from the `User` field of the config for the intermediate image generated from the application of the previous Dockerfile.
+
+#### Ability to rebase
+
+When extending the run image, a Dockerfile MAY indicate that it preserves the ability to rebase by including the following instruction:
+
+```bash
+LABEL io.buildpacks.rebasable=true
+```
+
+For the final image to be rebasable, the provided run image and all applied Dockerfiles MUST indicate rebasability.
```

Comment:
What will happen in case this label is set to true? What are the cases when an extension ran and the output image is still rebaseable? 

One other thing I am unsure about is how we will uniquely identify a "family" of base images that can be rebased with one another. 

I know we thought about replacing stack id with some other identifier that indicates the "family" of base images - how does that fit in here?

Reply:
The whole rebase picture with extensions is still unclear to me. It would be nice if we can have some actual examples of cases where/how this can work with platforms like pack or kpack.

Reply:
I'm thinking / hoping that an extension can still set rebaseable=true, where it modifies the build image (say to install runtimes, tools to build etc) .. but limits the runtime changes to just using 'FROM' to switch to a different run image.. I think this would still qualify to be rebaseable, as future run images can still be swapped in.. provided they still contain the required dependencies that were met by the substituted run image. In theory this allows for the existence of a generic builder image that tailors itself to install runtimes/tools etc, working in conjunction with a set of run images to cover the most common scenarios.. and maybe, also in conjunction with a generic (non rebaseable) run image to cover the edge cases where the common image was insufficient.

Reply:
I wonder, in the case where a Dockerfile is applied, but all it does is switch the base image (using a `FROM` instruction):
* Should we publish a new "extended" run image (the `<target-image>` argument provided to the `extender`)
  * Opinion: no, that seems confusing
* Should we run `genpkgs`?
  * Opinion: no, but we should make an effort to locate an sbom for the new run image (see https://github.com/buildpacks/rfcs/pull/211)